### PR TITLE
add_transpose_flatten_concat_fuse 

### PR DESCRIPTION
**Pull request by NHZlX**
1. add_transpose_flatten_concat_fuse 
In detect models like ssd, there exist patten of "transpose-> flatten ->concat ", we can fuse those op into one fused op to speed up inference time.
fullsizerender

2. fix elementwise mul op converter bug.

### CODE REVIEW

File: paddle/fluid/framework/ir/transpose_flatten_concat_fuse_pass.cc [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+#include "paddle/fluid/framework/ir/graph_viz_pass.h"
+#include "paddle/fluid/framework/ir/node.h"
+#include "paddle/fluid/framework/ir/transpose_flatten_concat_fuse_pass.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+template <int times>
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/transpose_flatten_concat_fuse_pass.cc:26]: what is `times` mean?
Need some comments here.

File: paddle/fluid/framework/ir/transpose_flatten_concat_fuse_pass.cc [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                         ->assert_is_op_input("transpose2", "X")
+                         ->AsInput();
+  }
+
+  patterns::TransposeFlattenConcat pattern(gpd.mutable_pattern(), pattern_name);
+  pattern(input_nodes, times);
+
+  auto handler = [&](const GraphPatternDetector::subgraph_t &subgraph,
+                     Graph *g) {
+    Node *nodes[times * 5];
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/transpose_flatten_concat_fuse_pass.cc]: Where is `5` comes from? No magic number here.

File: paddle/fluid/framework/ir/graph_pattern_detector.h [hunk 3]
```diff
@@ -766,6 +766,21 @@ struct ConvAffineChannel : public PatternBase {
   PATTERN_DECL_NODE(ac_out);  // Out
 };
 
+struct TransposeFlattenConcat : public PatternBase {
+  TransposeFlattenConcat(PDPattern* pattern, const std::string& name_scope)
+      : PatternBase(pattern, name_scope, "transpose_flatten_concat") {}
+
+  PDNode* operator()(PDNode* conv_input[], int times);
+
+  std::string GetNodeName(std::string op_type) {
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/graph_pattern_detector.h]: `const std::string&`

File: paddle/fluid/framework/ir/graph_pattern_detector.cc [hunk 4]
```diff
@@ -1306,6 +1306,55 @@ PDNode *patterns::ConvAffineChannel::operator()(
   return ac_out_var;
 }
 
+PDNode *patterns::TransposeFlattenConcat::operator()(PDNode *conv_in[],
+                                                     int times) {
+  // {trans, trans_out, flatten, flatten_out} * 4
+  PDNode *nodes[times * 4];
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/graph_pattern_detector.cc]: where is `4` comes from, no magic digit here.
Use a const variable instead, and add some comments.

File: paddle/fluid/framework/ir/CMakeLists.txt [hunk 5]
```diff
@@ -46,6 +46,12 @@ pass_library(conv_elementwise_add_act_fuse_pass inference)
 pass_library(conv_elementwise_add2_act_fuse_pass inference)
 pass_library(conv_elementwise_add_fuse_pass inference)
 pass_library(conv_affine_channel_fuse_pass inference)
+pass_library(transpose_flatten_concat_fuse_pass inference)
+
+foreach (index RANGE 3 6)
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/CMakeLists.txt:56]: Some comments here to describe this scenario about the special codes here.

File: paddle/fluid/framework/ir/graph_pattern_detector.cc [hunk 6]
```diff
@@ -1306,6 +1306,55 @@ PDNode *patterns::ConvAffineChannel::operator()(
...
+                       ->assert_op_has_n_inputs("concat", times);
+  auto concat_out = pattern->NewNode(GetNodeName("concat_out"))
+                        ->assert_is_op_output("concat")
+                        ->AsOutput();
+
+  std::vector<PDNode *> flatten_outs;
+  for (int i = 0; i < times; i++) {
+    conv_in[i]->AsInput();
+    // trans
+    nodes[i * 4]->LinksFrom({conv_in[i]});
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/graph_pattern_detector.cc]: `+1`, `+2` not intuitive, and there are used more than one time, how about

```c++
const int kNumFields = 4;
const int kTransOffset = 0;
const int kTransOutOffset = 1;
const int kFlattenOffset = 2;
const int kFlattenOutOffset = 3;

nodes[kNumFields+kTransOffset]->LinksFrom(...
```

File: paddle/fluid/framework/ir/transpose_flatten_concat_fuse_pass.cc [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+    // Create a new node for the fused op.
+    auto *new_conv_op = graph->CreateOpNode(&new_op_desc);
+
+    std::unordered_set<const Node *> delete_nodes;
+
+    for (int i = 0; i < times; i++) {
+      nodes[i * 5]->outputs.push_back(new_conv_op);
+      new_conv_op->inputs.push_back(nodes[i * 5]);
+      delete_nodes.insert(nodes[i * 5 + 1]);
```
Comment (Superjomn) [inline paddle/fluid/framework/ir/transpose_flatten_concat_fuse_pass.cc]: same here.

**Review by panyx0718 — APPROVED** [review]
LG const_cast

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**